website/src/_posts/2022-09-3.0.md

```diff
@@ -0,0 +1,146 @@
+
+You can also checkout how we migrated the Robodog example ourselves in this [commit](https://github.com/transloadit/uppy/commit/089aaed615c77bafaf905e291b6b4e82aaeb2f6f).
+
+```js
+// Import just the plugins you need
+import Uppy from '@uppy/core'
+import Dashboard from '@uppy/dashboard'
+import RemoteSources from '@uppy/remote-sources'
+import Webcam from '@uppy/webcam'
+import ScreenCapture from '@uppy/screen-capture'
+import ImageEditor from '@uppy/image-editor'
+import Audio from '@uppy/audio'
+import Transloadit, { COMPANION_URL, COMPANION_ALLOWED_HOSTS } from '@uppy/transloadit'
+
+// Import their styles
+import '@uppy/core/dist/style.css'
+import '@uppy/dashboard/dist/style.css'
+import '@uppy/webcam/dist/style.css'
+import '@uppy/audio/dist/style.css'
+import '@uppy/screen-capture/dist/style.css'
+import '@uppy/image-editor/dist/style.css'
+
+// Initialize Uppy
+const uppy = new Uppy()
+  .use(Dashboard, { inline: true, target: '#app' })
+  .use(RemoteSources, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
+  .use(Webcam, { target: Dashboard })
+  .use(Audio, { target: Dashboard })
+  .use(ScreenCapture, { target: Dashboard })
+  .use(ImageEditor, { target: Dashboard })
+  .use(Transloadit, {
+    async getAssemblyOptions (file) {
+      // This is where you configure your auth key, auth secret, and template ID
+      // https://uppy.io/docs/transloadit/#getAssemblyOptions-file
+      //
+      // It is important to set the secret in production:
+      // https://transloadit.com/docs/topics/signature-authentication/
+      const response = await fetch('/some-endpoint')
+      return response.json()
+    },
+  })
+```
+
+## Companion 4.0
+
+### Streaming upload
+
+Streaming upload can now be enabled in Companion with Tus. This gives greatly improved upload speeds when using remote Providers and allows uploading potentially hundreds of gigabytes without needing a large server storage.
+
+With streaming upload disabled (default), the whole file will be downloaded first, and only when the download has completely finished would the upload start.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+const tempFilePath = 'tmpfile'
+const fileStream = fs.createWriteStream(tempFilePath)
+await pipeline(providerStream, fileStream)
+await tus.uploadFile(tempFilePath)
+```
+
+When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+await tus.uploadStream(providerStream)
+```
```

Comment:
These code snippets show internals which AFAIK are not valuable or necessarily interesting for a blog post reader.

Reply:
I thought so too, but @arturi suggested that more details are better, because the readers are devs, see https://github.com/transloadit/uppy/issues/4042#issuecomment-1228423538

Reply:
I understand, but more details don't mean any details. It's pseudo internal code, I just don't think it adds anything here. If we want more details, the way to go is adding stats about the speed improvements.